### PR TITLE
[Ready]Refactoring smoke clouds

### PR DESCRIPTION
**Pull request by Ghommie**
## About The Pull Request
Atomizing the smoke system refactor from the rest of that PR. I previously successfully tested this locally, but reviews are very welcomed.

## Why It's Good For The Game
Updated the smoke system, less old code.

## Changelog
:cl:
refactor: Refactored the smoke effects system.
fix: Chemical smoke clouds shouldn't magically multiply reagents contained to affected mobs as much as before.
tweak: Reworked xeno acid/neuro smokes effects and damage calculation a bit.
tweak: tweaked some effects on other kind of smokes too.
fix: Fixed heat increment from certain reagents not capping properly.
fix: Fixed bio permeability coefficients not working as intented.
/:cl:

### CODE REVIEW

**Comment by DominikPanic** [comment]
Boiler smoke kills in under 2 seconds and melts barricades and razorwire with a single glob.

File: code/__HELPERS/game.dm [hunk 1]
```diff
@@ -406,7 +417,33 @@ datum/projectile_data
...
+	for(i in 1 to contents)
+		weight[i] /= listsum
+
+	//mix them
+	var/mixedcolor = 0
+	for(i in 1 to contents)
+		mixedcolor += weight[i]*color[i]
+	mixedcolor = round(mixedcolor)
+
+	//until someone writes a formal proof for this algorithm, let's keep this in
```
Comment (DominikPanic) [inline code/__HELPERS/game.dm]: What's the purpose of this?

File: code/datums/spell.dm [hunk 2]
```diff
@@ -171,15 +171,15 @@ var/list/spells = typesof(/obj/effect/proc_holder/spell) //needed for the badmin
...
-				var/datum/effect_system/smoke_spread/smoke = new /datum/effect_system/smoke_spread()
-				smoke.set_up(smoke_amt, 0, location) //no idea what the 0 is
-				smoke.start()
-			else if(smoke_spread == 2)
-				var/datum/effect_system/smoke_spread/bad/smoke = new /datum/effect_system/smoke_spread/bad()
-				smoke.set_up(smoke_amt, 0, location) //no idea what the 0 is
-				smoke.start()
+		var/datum/effect_system/smoke_spread/smoke
+		switch(smoke_spread)
+			if(1)
```
Comment (DominikPanic) [inline code/datums/spell.dm]: Magic numbers, could use defines.

File: code/game/mecha/combat/marauder.dm [hunk 3]
```diff
@@ -53,7 +53,7 @@
 	ME.attach(src)
 	ME = new /obj/item/mecha_parts/mecha_equipment/antiproj_armor_booster(src)
 	ME.attach(src)
-	src.smoke_system.set_up(3, 0, src)
+	src.smoke_system.set_up(3, src)
```
Comment (DominikPanic) [inline code/game/mecha/combat/marauder.dm]: src is implied

File: code/game/objects/effects/effect_system/smoke.dm [hunk 4]
```diff
@@ -7,198 +7,237 @@
...
-		var/obj/effect/particle_effect/smoke/foundsmoke = locate() in T //Don't spread smoke where there's already smoke!
-		if(foundsmoke)
+		smoke_mob(L)
+
+/obj/effect/particle_effect/smoke/proc/spread_smoke()
+	var/turf/t_loc = get_turf(src)
+	if(!t_loc)
+		return
+	var/list/newsmokes = list()
+	for(var/turf/T in get_adjacent_open_turfs(src))
```
Comment (DominikPanic) [inline code/game/objects/effects/effect_system/smoke.dm]: Is there a need to typecheck this?

File: code/game/objects/effects/effect_system/smoke.dm [hunk 5]
```diff
@@ -273,139 +278,99 @@
...
+/obj/effect/particle_effect/smoke/xeno/copy_stats(obj/effect/particle_effect/smoke/xeno/parent)
+	strength = parent.strength
+	return ..()
+
+/obj/effect/particle_effect/smoke/xeno/smoke_mob(mob/living/carbon/C)
+	if(lifetime < 1 || !istype(C))
+		return FALSE
+	if(C.stat == DEAD)
+		return FALSE
+	if(isxeno(C) || (isyautja(C) && prob(75)))
```
Comment (DominikPanic) [inline code/game/objects/effects/effect_system/smoke.dm]: Not sure if we need to snowflake preds into new code

File: code/game/objects/effects/effect_system/smoke.dm [hunk 6]
```diff
@@ -419,11 +384,119 @@ datum/effect_system/smoke_spread/tactical
...
+/datum/effect_system/smoke_spread/chem
+	var/obj/chemholder
+	smoke_type = /obj/effect/particle_effect/smoke/chem
+
+/datum/effect_system/smoke_spread/chem/New()
+	. = ..()
+	chemholder = new()
+	chemholder.create_reagents(500)
+
+/datum/effect_system/smoke_spread/chem/set_up(datum/reagents/carry = null, radius = 1, loca, smoke_time, silent = FALSE)
```
Comment (DominikPanic) [inline code/game/objects/effects/effect_system/smoke.dm]:
```suggestion
/datum/effect_system/smoke_spread/chem/set_up(datum/reagents/carry, radius = 1, loca, smoke_time, silent = FALSE)
```
No default value is equal to null

File: code/game/objects/effects/effect_system/smoke.dm [hunk 7]
```diff
@@ -419,11 +384,119 @@ datum/effect_system/smoke_spread/tactical
...
+		if(contained)
+			contained = "\[[contained]\]"
+
+		var/where = "[AREACOORD(location)]"
+		if(carry.my_atom.fingerprintslast)
+			var/mob/M = get_mob_by_key(carry.my_atom.fingerprintslast)
+			var/more = ""
+			if(M)
+				more = "[ADMIN_LOOKUPFLW(M)] "
+			message_admins("Smoke: ([ADMIN_VERBOSEJMP(location)])[contained]. Key: [more ? more : carry.my_atom.fingerprintslast].")
```
Comment (DominikPanic) [inline code/game/objects/effects/effect_system/smoke.dm]: Could you somehow turn the key into `ADMIN_TPMONTY`? Would make muh adminning easier, not a requirement though.

File: code/__HELPERS/game.dm (hunk 1 above)
Comment (Ghommie) [reply code/__HELPERS/game.dm]: copypasta from /tg/station, didn't really give an in-deep look on it, yet.

Comment (Ghommie) [reply code/__HELPERS/game.dm]: actually, it's cut and pasted from this repo's Chemistry-Colours.dm.

File: code/game/objects/effects/effect_system/smoke.dm (hunk 5 above)
Comment (Ghommie) [reply code/game/objects/effects/effect_system/smoke.dm]: I feel like retaining a bit of legacy here until we get to an overhaul or something going, alas.

Comment (DominikPanic) [reply code/game/objects/effects/effect_system/smoke.dm]: Tbh most snowflake predator things are slowly being removed (they are no longer in prefs, whitelist has been borked for a long time, they're only left in the admin menu at this point)

Comment (Ghommie) [reply code/game/objects/effects/effect_system/smoke.dm]: fine then.

File: code/__HELPERS/game.dm [hunk 8]
```diff
@@ -22,6 +22,17 @@
 		return null
 	return format_text ? format_text(A.name) : A.name
 
+/proc/get_open_turf_in_dir(atom/center, dir)
+	var/turf/open/T = get_ranged_target_turf(center, dir, 1)
```
Comment (Rohesie) [inline code/__HELPERS/game.dm]:
```suggestion
	var/turf/open/T = get_step(center, dir)
```

File: code/modules/reagents/Chemistry-Colours.dm [hunk 9]
```diff
@@ -1,66 +1,21 @@
...
+	var/vol_temp
 
-	//fill the list of weights
-	for(i=1; i<=contents; i++)
-		var/datum/reagent/re = reagent_list[i]
-		var/reagentweight = re.volume
-		if(istype(re, /datum/reagent/paint))
-			reagentweight *= 20 //Paint colours a mixture twenty times as much
-		weight[i] = reagentweight
+	for(var/datum/reagent/R in reagent_list)
```
Comment (DominikPanic) [inline code/modules/reagents/Chemistry-Colours.dm]: Could optimize it by not typecasting

File: code/game/objects/effects/effect_system/smoke.dm [hunk 10]
```diff
@@ -7,208 +7,243 @@
...
 	return ..()
 
+/obj/effect/particle_effect/smoke/tactical/kill_smoke()
+	apply_smoke_effect(get_turf(src))
+	return ..()
 
-/obj/effect/particle_effect/smoke/tactical/affect(var/mob/living/M)
-	cloak_smoke_act(M)
+/obj/effect/particle_effect/smoke/tactical/smoke_mob(mob/living/M)
+	if(istype(M))
```
Comment (DominikPanic) [inline code/game/objects/effects/effect_system/smoke.dm]: Early return, but ideally this shouldn't be called on non-living mobs

File: code/game/objects/effects/effect_system/smoke.dm [hunk 11]
```diff
@@ -429,11 +379,116 @@ datum/effect_system/smoke_spread/tactical
...
+
+/////////////////////////////////////////////
+// Chem smoke
+/////////////////////////////////////////////
+/obj/effect/particle_effect/smoke/chem
+	lifetime = 10
+	var/list/smoked_mobs
+
+/obj/effect/particle_effect/smoke/chem/Destroy()
+	if(length(smoked_mobs) && alpha) //so the whole cloud won't stop working somehow
```
Comment (DominikPanic) [inline code/game/objects/effects/effect_system/smoke.dm]: You are not calling parent on Destroy

File: code/_globalvars/bitfields.dm [hunk 12]
```diff
@@ -189,7 +189,7 @@ GLOBAL_LIST_INIT(bitfields, list(
 		"HAND_LEFT" = HAND_LEFT,
 		"HAND_RIGHT" = HAND_RIGHT
 		),
-	"flags_armor_protection" = list(
+	"body_parts_covered" = list(
```
Comment (spookydonut) [inline code/_globalvars/bitfields.dm]: if this is just a variable name change then atomise this out of the PR so its not fucking 81 files

File: code/datums/diseases/mob_procs.dm [hunk 13]
```diff
@@ -146,15 +146,15 @@
 				Cl = SLOT_W_UNIFORM
 				. = prob((Cl.permeability_coefficient*100) - 1)
 		if(3)
-			if(isobj(wear_suit) && wear_suit.flags_armor_protection&HANDS)
+			if(isobj(wear_suit) && wear_suit.body_parts_covered&HANDS)
```
Comment (spookydonut) [inline code/datums/diseases/mob_procs.dm]: eg this
```suggestion
			if(isobj(wear_suit) && wear_suit.body_parts_covered & HANDS)
```

File: code/game/objects/effects/effect_system/smoke.dm [hunk 14]
```diff
@@ -7,208 +7,242 @@
...
 	STOP_PROCESSING(SSobj, src)
-	. =..()
+	for(var/obj/effect/particle_effect/smoke/C in current_cloud)
+		C.current_cloud -= src
+	return ..()
 
+/obj/effect/particle_effect/smoke/proc/kill_smoke()
+	STOP_PROCESSING(SSobj, src)
+	INVOKE_ASYNC(src, .proc/fade_out)
+	QDEL_IN(src, 10)
```
Comment (spookydonut) [inline code/game/objects/effects/effect_system/smoke.dm]: uh no, do it at the end of the fade_out proc

File: code/game/objects/effects/effect_system/smoke.dm [hunk 15]
```diff
@@ -7,208 +7,242 @@
...
+	lifetime = 8
 
-/obj/effect/particle_effect/smoke/bad/Move()
-	..()
-	for(var/mob/living/carbon/M in get_turf(src))
-		affect(M)
+/obj/effect/particle_effect/smoke/bad/effect_inhale(mob/living/carbon/C)
+	if(prob(30))
+		C.drop_held_item()
+	C.adjustOxyLoss(1)
```
Comment (spookydonut) [inline code/game/objects/effects/effect_system/smoke.dm]: scream this should be handled by a carbon proc not a smoke proc.

File: code/game/objects/effects/effect_system/smoke.dm [hunk 16]
```diff
@@ -7,208 +7,242 @@
...
-			M.drop_held_item()
-		M.adjustOxyLoss(1)
-		if(M.coughedtime != 1)
-			M.coughedtime = 1
-			if(ishuman(M)) //Humans only to avoid issues
-				M.emote("cough")
-			spawn(20)
-				M.coughedtime = 0
+/obj/effect/particle_effect/smoke/bad/CanPass(atom/movable/mover, turf/target)
+	if(istype(mover, /obj/item/projectile/beam))
```
Comment (spookydonut) [inline code/game/objects/effects/effect_system/smoke.dm]:
```suggestion
	if(istype(mover, /obj/item/projectile))
		var/obj/item/projectile/P = mover
		P.handle_smoke_traversal(SMOKE_VERY_DENSE)
```

File: code/game/objects/effects/effect_system/smoke.dm [hunk 17]
```diff
@@ -7,208 +7,242 @@
...
-	M.drop_held_item()
-	M:sleeping += 1
-	if(M.coughedtime != 1)
-		M.coughedtime = 1
-		if(ishuman(M)) //Humans only to avoid issues
-			M.emote("cough")
-		spawn(20)
-			M.coughedtime = 0
+/obj/effect/particle_effect/smoke/sleepy/effect_inhale(mob/living/carbon/C)
+	C.Sleeping(1)
```
Comment (spookydonut) [inline code/game/objects/effects/effect_system/smoke.dm]: same here

**Comment by DominikPanic** [comment]
Have you tested it locally?

File: code/modules/clothing/clothing.dm [hunk 18]
```diff
@@ -221,7 +221,7 @@
 /obj/item/clothing/mask
 	name = "mask"
 	icon = 'icons/obj/clothing/masks.dmi'
-	flags_armor_protection = HEAD
+	flags_armor_protection = 0
```
Comment (DominikPanic) [inline code/modules/clothing/clothing.dm]:
```suggestion
	flags_armor_protection = NOFLAGS
```

File: code/game/objects/items.dm [hunk 19]
```diff
@@ -33,7 +33,7 @@
 
 	var/obj/item/master = null
 
-	var/flags_armor_protection = NOFLAGS //see setup.dm for appropriate bit flags
+	var/flags_armor_protection = 0 //see setup.dm for appropriate bit flags
```
Comment (DominikPanic) [inline code/game/objects/items.dm]:
```suggestion
	var/flags_armor_protection = NOFLAGS //see setup.dm for appropriate bit flags
```

**Comment by Rohesie** [comment]
Please let us know when you test this locally, after fixing conflicts.

**Comment by Ghommie** [comment]
> 
> 
> Please let us know when you test this locally, after fixing conflicts.

Yes, it has been tested, I forgot to tell you before.

**Comment by Rohesie** [comment]
Boiler globs sometimes work, sometimes don't.

**Comment by Ghommie** [comment]
> 
> 
> Boiler globs sometimes work, sometimes don't.
Should be fixed now. It seemed to be happening around xeno hosts who were either buckled to a next or dead.

File: code/__HELPERS/game.dm [hunk 20]
```diff
@@ -22,6 +22,17 @@
 		return null
 	return format_text ? format_text(A.name) : A.name
 
+/proc/get_open_turf_in_dir(atom/center, dir)
+	var/turf/open/T = get_step(center, dir)
+	if(istype(T))
+		return T
+
+/proc/get_adjacent_open_turfs(atom/center)
+	. = list(get_open_turf_in_dir(center, NORTH),
```
Comment (Rohesie) [inline code/__HELPERS/game.dm]:
```suggestion
. = list()
for(var/i in GLOB.cardinals)
	var/turf/open/T = get_step(center, i)
	if(!istype(T))
		continue
	. += T
```

Comment (Rohesie) [reply code/__HELPERS/game.dm]: One less proc. Though this is optional, I suppose, the difference is small.

File: code/__HELPERS/game.dm [hunk 21]
```diff
@@ -22,6 +22,18 @@
 		return null
 	return format_text ? format_text(A.name) : A.name
 
+/proc/get_open_turf_in_dir(atom/center, dir)
```
Comment (Rohesie) [inline code/__HELPERS/game.dm]: This proc is no longer needed.

**Comment by DominikPanic** [comment]
This has been tested without further issues found.

**Pull request merged**